Full height swipe-able menus

featheredtoast opened the pull request:
Add full height menu on mobile, animated slide in menus, and swipe-to-dismiss menu gestures. Android additionally gets swipe-to-slide-in gestures for menus. (iOS does not respond to this as it's reserved for forward/back motions. Dismissal swipes still work from the center of the screen).

Adds swipe support for iOS.

coding-horror commented:
https://media.giphy.com/media/asHT7eh4AwG9G/200.gif

jjaffeux commented on app/assets/javascripts/discourse/components/site-header.js.es6:
```diff
@@ -23,6 +28,135 @@ const SiteHeaderComponent = MountWidget.extend(Docking, {
+        this.eventDispatched("dom:clean", "header");
+        this.set("panMenuOffset", 0);
+      } else if (offset <= 0) {
+        $panel.css("right", "");
```
```
$panel.css({ right: "", left: "" });
```

jjaffeux commented on app/assets/javascripts/discourse/components/site-header.js.es6:
```diff
@@ -180,6 +354,18 @@ const SiteHeaderComponent = MountWidget.extend(Docking, {
+          $headerCloak.removeClass("animate");
+        }, 200);
+      }
+      $panel.css("right", "");
```
$panel.css({ right: "", left: "" });

jjaffeux commented on app/assets/javascripts/discourse/widgets/header.js.es6:
```diff
@@ -416,10 +428,32 @@ export default createWidget("header", {
+  toggleBodyScrolling(bool) {
+    if (!this.site.mobileView) return;
+    if (bool) {
+      document.body.addEventListener('touchmove', this.preventDefault, { passive: false });
```
```suggestion
      document.body.addEventListener("touchmove", this.preventDefault, { passive: false });
```

jjaffeux commented on app/assets/javascripts/discourse/widgets/header.js.es6:
```diff
@@ -416,10 +428,32 @@ export default createWidget("header", {
+    if (bool) {
+      document.body.addEventListener('touchmove', this.preventDefault, { passive: false });
+    } else {
+      document.body.removeEventListener('touchmove', this.preventDefault, { passive: false });
```
```suggestion
      document.body.removeEventListener("touchmove", this.preventDefault, { passive: false });
```

jjaffeux commented on app/assets/javascripts/discourse/widgets/header.js.es6:
```diff
@@ -416,10 +428,32 @@ export default createWidget("header", {
+    let height = window.innerHeight
+        ? window.innerHeight
+        : $(window).height();
+    if (!$(e.target).parents('.menu-panel').length ||
```
```suggestion
    if (!$(e.target).parents(".menu-panel").length ||
```

jjaffeux commented on app/assets/javascripts/discourse/widgets/header.js.es6:
```diff
@@ -416,10 +428,32 @@ export default createWidget("header", {
+        ? window.innerHeight
+        : $(window).height();
+    if (!$(e.target).parents('.menu-panel').length ||
+        $('.menu-panel .panel-body-contents').height() <= height) {
```
```suggestion
        $(".menu-panel .panel-body-contents").height() <= height) {
```

jjaffeux commented:
If you can’t add tests to this code (which I understand given the mobile and gesture nature of the feature) please add more comments and maybe more explicit code. ATM, I feel like I wouldn’t want to have to modify this in the future, it's a lot of special cases with very few explanations.

jjaffeux commented on app/assets/javascripts/discourse/widgets/header.js.es6:
```diff
@@ -416,10 +428,32 @@ export default createWidget("header", {
+
+  preventDefault(e) {
+    // prevent all scrollin on menu panels, except on overflow
+    let height = window.innerHeight
```
```suggestion
    const height = window.innerHeight
```

jjaffeux commented on app/assets/javascripts/discourse/widgets/user-menu.js.es6:
```diff
@@ -194,7 +194,23 @@ export default createWidget("user-menu", {
+      this.sendWidgetAction("toggleUserMenu");
+    } else {
+      const $window = $(window);
+      const windowWidth = parseInt($window.width());
```
It's good practice to always use the radix, even if it should never be an issue here.

```suggestion
      const windowWidth = parseInt($window.width(), 10);
```

jjaffeux commented on app/assets/javascripts/discourse/widgets/hamburger-menu.js.es6:
```diff
@@ -323,7 +323,24 @@ export default createWidget("hamburger-menu", {
+      this.sendWidgetAction("toggleHamburger");
+    } else {
+      const $window = $(window);
+      const windowWidth = parseInt($window.width());
```
```suggestion
      const windowWidth = parseInt($window.width(), 10);
```

jjaffeux commented on app/assets/javascripts/discourse/components/site-header.js.es6:
```diff
@@ -160,16 +332,18 @@ const SiteHeaderComponent = MountWidget.extend(Docking, {
+        let height = window.innerHeight
+          ? window.innerHeight
+          : $(window).height();
+        const winHeight = height - 16;
```
can you clarify the meaning of 16 here ?

jjaffeux commented on app/assets/javascripts/discourse/components/site-header.js.es6:
```diff
@@ -23,6 +28,135 @@ const SiteHeaderComponent = MountWidget.extend(Docking, {
+  _panOpenClose(offset, velocity, direction) {
+    const $window = $(window);
+    const windowWidth = parseInt($window.width());
+    const $menuPanels = $(".menu-panel");
```
used only once

eviltrout commented on app/assets/javascripts/discourse/components/site-header.js.es6:
```diff
@@ -23,6 +28,135 @@ const SiteHeaderComponent = MountWidget.extend(Docking, {
+        $panel.css("left", "");
+        this.set("panMenuOffset", 0);
+      } else {
+        Ember.run.later(
```
I wonder if this event should check if one is already scheduled and clear it first? 20ms is pretty fast but is there any chance two could be enqueued at once? I would recommend storing the timer in a private variable and calling Ember.run.cancel before.

You should also clear the timeout in `willDestroyElement`

eviltrout commented on app/assets/javascripts/discourse/components/site-header.js.es6:
```diff
@@ -23,6 +28,135 @@ const SiteHeaderComponent = MountWidget.extend(Docking, {
+      e.originalEvent.preventDefault();
+      this.set("isPanning", true);
+    } else if (
+      center.x < 30 &&
```
There are a couple more constants here, 30 and 300 it seems.

eviltrout commented on app/assets/javascripts/discourse/components/site-header.js.es6:
```diff
@@ -74,6 +209,14 @@ const SiteHeaderComponent = MountWidget.extend(Docking, {
       }
     });
+
+    if (this.site.mobileView) {
```
Would it be better to check the capabilities object for touch? Some touch devices like ipad are not in mobileView. `this.capabilities.touch`

eviltrout commented on app/assets/javascripts/discourse/components/site-header.js.es6:
```diff
@@ -180,6 +354,18 @@ const SiteHeaderComponent = MountWidget.extend(Docking, {
+      if (animate) {
+        $panel.addClass("animate");
+        $headerCloak.addClass("animate");
+        Ember.run.later(() => {
```
This is another event you will want to make sure only happens once and is cleared on destroy (see my earlier comments.)

eviltrout commented on app/assets/javascripts/discourse/mixins/pan-events.js.es6:
```diff
@@ -7,11 +11,23 @@ export default Ember.Mixin.create({
-        .on("pointermove", e => this._panMove(e))
-        .on("pointerup", e => this._panMove(e))
-        .on("pointercancel", e => this._panMove(e));
+      if ("onpointerdown" in document.documentElement) {
```
This could use `this.capabilities.touch` too right? If not, should it be added to capabilities?

featheredtoast commented:
Thanks for the concerns/feedback - I've implemented most of the feedback given so far.

Two points about the capacities, and rationale behind mobile only - tldr, it'd be nice to get to supporting all touch devices at some point but I'm focusing/optimizing for mobile only here:

onpointerdown seems to be a newer touch API, not (yet) supported in iOS/webkit - I wrote against that API before I realized ontouch is the only thing I've got on iOS. Probably not worth adding to capacities, I added this as a shim to support both APIs, rather than rewrite against ontouch only.

I also wanted this work to be incremental, which is why I opted to not support anything other than mobile views (for now). While we could add support for ipads and tablets, that would be a secondary goal for me. The work as is also includes masking the side of a screen for a menu, and optimizes touch events for mobile specifically where unmasked content is not useful at all.

eviltrout commented on app/assets/javascripts/discourse/components/site-header.js.es6:
```diff
@@ -23,6 +30,146 @@ const SiteHeaderComponent = MountWidget.extend(Docking, {
+
+  panStart(e) {
+    //android supports pulling in from the screen edges
+    const screenEdgeMargin = 30;
```
I would recommend pulling these out and putting them at the top of the file like `SCREN_EDGE_MARGIN = 30`. It's nice to collect those kinds of things up top in case you want to export them later.

eviltrout commented on app/assets/javascripts/discourse/components/topic-navigation.js.es6:
```diff
@@ -136,10 +136,11 @@ export default Ember.Component.extend(PanEvents, {
 
   _shouldPanClose(e) {
-    return (e.deltaY > 200 && e.velocityY > -0.15) || e.velocityY > 0.15;
+    return (e.deltaY > 200 && e.velocityY > -0.10) || e.velocityY > 0.10;
```
no constants for this stuff?

featheredtoast replied:
Oops! missed those

featheredtoast commented:
Refactored the shared ones into the pan events mixin to be used in both site-header and topic-navigation.

Additionally, I've refactored topic-navigation to better match the feedback I got from Joffrey.

davidtaylorhq commented:
@featheredtoast you may already be aware, but for me the slide-in menus happen on Desktop when my browser is less than about 1200px wide. Is that deliberate?

featheredtoast commented:
@davidtaylorhq Was pointed out in chat on Tuesday as well I believe. I've changed the behavior so it's only switched for mobile views only.
